File: yh_service/src/yh_client.cpp
#include "ros/ros.h"
#include "yh_service/YhSrv.h" //서비스 헤더 파일
                            // 빌드시 생성
#include <cstdlib> //atoi 함수 사용을 위한 라이브러리 : 문자열을 정수로 만들어주는 함수

int main(int argc, char** argv){

    ros::init(argc, argv, "yh_client"); //노드 이름 초기화

    //{"yh_client", "a", "b"}
    if(argc != 3){ //argc : 몇개가 들어왔는지, argv : 뭐가 들어왔는지
        ROS_INFO("command: rosrun yh_service yh_client arg1 arg2");
        ROS_INFO("arg1, arg2 : int32 number");
        return 1;
    }

    ros::NodeHandle nh;

    // 서비스 클라이언트 (yh_client)를 선언한다.
    // 서비스 이름은 (multiply)이고 패키지 (yh_service)의 (YhSrv) 서비스 파일을 사용한다.
    ros::ServiceClient yh_client = nh.serviceClient<yh_service::YhSrv>("multiply");

    yh_service::YhSrv srv; // 서비스 가져올 변수 선언

    // 노드 실행시 입력된 변수를 서비스 요청 값의 a,b에 저장한다.
    srv.request.a = atoi(argv[1]);
    srv.request.b = atoi(argv[2]);

    if(yh_client.call(srv)){ //정상적으로 통신이 되었을때 출력
        ROS_INFO("a:%d, b:%d", srv.request.a, srv.request.b);
        ROS_INFO("result: %d", srv.response.result);
    }
    else{
        ROS_ERROR("Failed to call service");
        return 1;
    }

    return 0;
}

File: yh_service/scripts/yh_client.py
#!/usr/bin/python3
#-*- coding:utf-8 -*-

import rospy
from yh_service.srv import YhSrv, YhSrvRequest
import sys # 프로그램 실행할때 넣어주는 숫자(a,b)를 사용하기 위해 import

# service에 요청하고 값이 오면 반환
def multiply_client():
    rospy.init_node("yh_client")
    
    #실행 시 들어오는 값이 3개가 아니면 : {노드이름, a,b}
    if len(sys.argv) != 3:
        rospy.loginfo("rosrun yh_service yh_client.py a b")
        rospy.loginfo("a, b : int32 number")
        sys.exit(1)

    # 서비스 클라이언트 선언
    # ServiceProxy : 클라이언트를 만드는 함수 
    yh_client = rospy.ServiceProxy("multiply", YhSrv)

    req = YhSrvRequest()
    req.a = int(sys.argv[1])
    req.b = int(sys.argv[2])

    try:
        res = yh_client(req)
        rospy.loginfo(f"a: {req.a}, b: {req.b} result: {res.result}")
    except rospy.ServiceException as e:
        rospy.logerr(f"Fail: {e}")

   


if __name__ =="__main__":
    multiply_client()